tests/components/slack/test_notify.py

```diff
@@ -29,84 +36,397 @@
+        )  # Changed to handle one upload per channel
+
+
+async def test_get_service_discovery_info(hass: HomeAssistant) -> None:
+    """Test async_get_service with discovery info."""
+    mock_client = AsyncMock()
+    discovery_info = {SLACK_DATA: {DATA_CLIENT: mock_client}}
+
+    service = await async_get_service(hass, {}, discovery_info)
+
+    assert service is not None
+    assert isinstance(service, SlackNotificationService)
+    assert service._client == mock_client
+    assert service._hass == hass
+    assert service._config == discovery_info
+
+
+async def test_get_service_no_discovery_info(hass: HomeAssistant) -> None:
+    """Test async_get_service without discovery info."""
+    service = await async_get_service(hass, {}, None)
+    assert service is None
+
+
+@pytest.mark.parametrize(
+    ("target", "expected"),
+    [
+        # Empty/None handling
+        ("", ("channel_name", "")),
+        # User IDs
+        ("U123ABC456", ("user_id", "U123ABC456")),
+        ("W123ABC456", ("user_id", "W123ABC456")),
+        ("UABCDEF123", ("user_id", "UABCDEF123")),
+        # Channel IDs
+        ("C123ABC456", ("channel_id", "C123ABC456")),
+        ("G123ABC456", ("channel_id", "G123ABC456")),
+        ("D123ABC456", ("channel_id", "D123ABC456")),
+        # Channel names
+        ("general", ("channel_name", "general")),
+        ("#general", ("channel_name", "general")),
+        ("my-channel", ("channel_name", "my-channel")),
+        ("#my-channel", ("channel_name", "my-channel")),
+        # Invalid/Edge cases
+        ("U12345", ("channel_name", "U12345")),  # Too short for user ID
+        ("C12345", ("channel_name", "C12345")),  # Too short for channel ID
+        ("X123ABC456", ("channel_name", "X123ABC456")),  # Invalid prefix
+        ("U123ABC456DEF", ("user_id", "U123ABC456DEF")),  # Extra long but valid
+        ("not-an-id", ("channel_name", "not-an-id")),
+    ],
+)
+def test_process_target(target: str, expected: tuple[str, str]) -> None:
+    """Test target processing for various input formats."""
+    assert _async_process_target(target) == expected
+
+
+def test_process_target_not_string() -> None:
+    """Test handling of non-string input (shouldn't happen in practice but good to test)."""
+    # Type ignore because we're intentionally testing with wrong type
+    result = _async_process_target(123)  # type: ignore[arg-type]
+    assert result == ("channel_name", 123)  # type: ignore[comparison-overlap]
+
+
+def test_process_target_strip_multiple_hashes() -> None:
+    """Test that multiple # characters are all stripped."""
+    assert _async_process_target("###general") == ("channel_name", "general")
+
+
+async def test_dm_channel_api_error(mock_client) -> None:
+    """Test handling of Slack API error when opening DM channel."""
+    mock_client.conversations_open.side_effect = SlackApiError("error", {"ok": False})
+    service = SlackNotificationService(None, mock_client, CONF_DATA)
```

Comment:
Usually we want to setup the service and test this through the normal APIs instead of using the object directly.